### ISSUE TITLE
sync modifies vendor.json

### ISSUE THREAD

**Issue by moorereason** [kardianos/govendor#233]
### Problem

When I run `govendor sync`, the `vendor.json` file is being updated (incorrectly, too).

### Expectation

I expect govendor to simply read `vendor.json` and download the packages from remotes that match what's already recorded in `vendor.json`.  I don't expect `vendor.json` to be written to.

### Reproducing

I'm working on `github.com/spf13/hugo`.  I have `github.com/spf13/cast` in my local GOPATH, but my local copy is two revisions behind `cast` master (I have [60e7a69](https://github.com/spf13/cast/tree/60e7a69a428e9ac1cf7e0c865fc2fe810d34363e)).  Another hugo team member [updated our vendoring](https://github.com/spf13/hugo/commit/7cdd230e1649a12ef9825a715727ae3bfc87c296#diff-bd290170e2912d3e8694db1a151066e5) to the latest revision of `cast` ([2580bc9](https://github.com/spf13/cast/tree/2580bc98dc0e62908119e4737030cc2fdfc45e4c)).  The `vendor.json` file committed to hugo master is correct.  So far so good.

When I run `govendor sync` locally (with my out-of-date `cast` package in GOPATH), govendor pulls down the correct package revisions into `vendor/`, but incorrectly calculates the checksum for the `cast` package using the contents of my out-of-date `cast` package.  In-so-doing, govendor decides to update the `vendor.json` as follows:

```diff
diff --git a/vendor/vendor.json b/vendor/vendor.json
index 65c3979..1a637b6 100644
--- a/vendor/vendor.json
+++ b/vendor/vendor.json
@@ -251,7 +251,7 @@
                        "revisionTime": "2016-09-19T21:01:14Z"
                },
                {
-                       "checksumSHA1": "LYlvAvaPyvxqvn5j3frHjr8dwCY=",
+                       "checksumSHA1": "M4qI7Ul0vf2uj3fF69DvRnwqfd0=",
                        "path": "github.com/spf13/cast",
                        "revision": "2580bc98dc0e62908119e4737030cc2fdfc45e4c",
                        "revisionTime": "2016-09-26T08:42:49Z"
```

Notice that the new checksum is the same checksum from prior to the vendor updates by my hugo team member.

So, to recap:

1. govendor should not modify `vendor.json` when running in sync mode.  I'd rather see an error in this instance.
1. govendor should not use a local GOPATH package to calculate checksums when running in sync mode (cf. #129).

**Comment by kardianos** [comment]
Thank you for the detailed report. I'm torn regarding govendor should write to the checksum during sync or not. I need more thought on that when i'm less tired. 

It does sound like there is at least one other bug present for the incorrect checksum to be calculated. I'll look into that. I appreciate your detail.

**Comment by bep** [comment]
> I'm torn regarding govendor should write to the checksum during sync or not.

In what situations isn't `newRevision == oldRevision && newChecksumSHA1 != oldChecksumSHA1` a `log.Fatal`?

**Comment by kevinburke** [comment]
From discussion in Slack, I triggered this by adding `vendor.json` on one Mac (with all dependencies already present in GOPATH), then running `govendor sync` on a different Mac (with some dependencies not present).

It's possible the case insensitivity of filenames on Macs is playing a part here.

**Comment by alexandrevicenzi** [comment]
There's an easy way to reproduce. Create a project on Linux with govendor and run `govendor sync` on Windows. It will change the checksum too.

Here's a list of dependencies that change:

```diff
diff --git a/vendor/vendor.json b/vendor/vendor.json
index 52876f7..f3ff68e 100644
--- a/vendor/vendor.json
+++ b/vendor/vendor.json
@@ -3,133 +3,133 @@
        "ignore": "test",
        "package": [
                {
-                       "checksumSHA1": "AKjqgiWov1IJWbnpCkjVEs6NunQ=",
+                       "checksumSHA1": "EoomAu2qBa6LBhPt9waqSEFraHY=",
                        "path": "github.com/bmizerany/pat",
                        "revision": "c068ca2f0aacee5ac3681d68e4d0a003b7d1fd2c",
                        "revisionTime": "2016-02-17T10:30:50Z"
                },
                {
-                       "checksumSHA1": "D8gUHzS23ynNPS4imXcAqGnfTck=",
+                       "checksumSHA1": "6MafhETnezZmyBZZiWvD27LRGRU=",
                        "path": "github.com/certifi/gocertifi",
                        "revision": "a61bf5eafa3aee233ec8043e9da052447e5463dd",
                        "revisionTime": "2016-09-26T11:54:48Z"
                },
                {
-                       "checksumSHA1": "dvabztWVQX8f6oMLRyv4dLH+TGY=",
+                       "checksumSHA1": "Lps/+JqMRVlrXgO/+cazYcfsglg=",
                        "path": "github.com/davecgh/go-spew/spew",
                        "revision": "346938d642f2ec3594ed81d874461961cd0faa76",
                        "revisionTime": "2016-10-29T20:57:26Z"
                },
                {
-                       "checksumSHA1": "0VI4tGLlLDXRQXOX0MBO0qZWFTQ=",
+                       "checksumSHA1": "NOF9uwOgOEB1WvLT+0plzWiLG9I=",
                        "path": "github.com/eventials/go-tus",
                        "revision": "6cd40d7bd582a425a905c475e4192bcbec59a142",
                        "revisionTime": "2016-12-27T12:19:01Z"
                },
                {
-                       "checksumSHA1": "ZTSxs7RNme9GNyGkDPGEH8ePecU=",
+                       "checksumSHA1": "qNK2bTbZmrlMJJEuqrdfv2dss/w=",
                        "path": "github.com/eventials/golog",
                        "revision": "97cfe58620d6c5d1127fcc7b30d278f6c10b6e77",
                        "revisionTime": "2016-12-20T19:23:23Z"
                },
                {
-                       "checksumSHA1": "s3s8aYF/nnL4kpTj7lvmPPeuIIw=",
+                       "checksumSHA1": "i2oy0pC035fbmqJmfA27T2nA/Kg=",
                        "path": "github.com/eventials/golog/console",
                        "revision": "97cfe58620d6c5d1127fcc7b30d278f6c10b6e77",
                        "revisionTime": "2016-12-20T19:23:23Z"
                },
                {
-                       "checksumSHA1": "47KX5/Rj/+nHjNoG28EwszCG3tc=",
+                       "checksumSHA1": "1c7Po7rC1Wwvahd/Ece30p7wSFI=",
                        "path": "github.com/eventials/golog/sentry",
                        "revision": "97cfe58620d6c5d1127fcc7b30d278f6c10b6e77",
                        "revisionTime": "2016-12-20T19:23:23Z"
                },
                {
-                       "checksumSHA1": "fut7sZrX10muj3RMf0xrprbRCLo=",
+                       "checksumSHA1": "e++IEWaIfIVH+zQC8v7cEY6KC+c=",
                        "path": "github.com/eventials/golog/stream",
                        "revision": "97cfe58620d6c5d1127fcc7b30d278f6c10b6e77",
                        "revisionTime": "2016-12-20T19:23:23Z"
                },
                {
-                       "checksumSHA1": "3ur+hHXnLZOIghIdCBBMx9pbP4E=",
+                       "checksumSHA1": "LhE3kzOIzQa1g+you90GUE8Z+pc=",
                        "path": "github.com/getsentry/raven-go",
                        "revision": "3f7439d3e74d88e21d196ba20eb61a5a958bc118",
                        "revisionTime": "2016-11-15T13:54:11Z"
                },
                {
-                       "checksumSHA1": "g+afVQQVopBLiLB5pFZp/8s6aBs=",
+                       "checksumSHA1": "kuwREeq736QUok7wyPzN9QsesTo=",
                        "path": "github.com/kardianos/osext",
                        "revision": "c2c54e542fb797ad986b31721e1baedf214ca413",
                        "revisionTime": "2016-08-11T00:15:26Z"
                },
                {
-                       "checksumSHA1": "GB91LURB6wOiLnOxoIIoufxqHPw=",
+                       "checksumSHA1": "mDwPA7gZCwzut3F9+nXADuavg18=",
                        "path": "github.com/kardianos/service",
                        "revision": "6d3a0ee7d3425d9d835debc51a0ca1ffa28f4893",
                        "revisionTime": "2016-11-19T21:06:48Z"
                },
                {
-                       "checksumSHA1": "LuFv4/jlrmFNnDb/5SCSEPAM9vU=",
+                       "checksumSHA1": "R3yk3OJzuq4MVfVGX5AhkkYv1ko=",
                        "path": "github.com/pmezard/go-difflib/difflib",
                        "revision": "792786c7400a136282c1664665ae0a8db921c6c2",
                        "revisionTime": "2016-01-10T10:55:54Z"
                },
                {
-                       "checksumSHA1": "K0crHygPTP42i1nLKWphSlvOQJw=",
+                       "checksumSHA1": "ofNmmhhJAxYL1Mh09iq+utdfHIM=",
                        "path": "github.com/stretchr/objx",
                        "revision": "1a9d0bb9f541897e62256577b352fdbc1fb4fd94",
                        "revisionTime": "2015-09-28T12:21:52Z"
                },
                {
-                       "checksumSHA1": "/kKwYqD04AkkDZFQhGwklKw3RFA=",
+                       "checksumSHA1": "fPjZ9M1PDFi51iQ0aXEjQG+qLqM=",
                        "path": "github.com/stretchr/testify",
                        "revision": "2402e8e7a02fc811447d11f881aa9746cdc57983",
                        "revisionTime": "2016-12-17T20:04:45Z"
                },
                {
-                       "checksumSHA1": "JXUVA1jky8ZX8w09p2t5KLs97Nc=",
+                       "checksumSHA1": "ziYrpyU5zo1q9+n1kaXqkR3kP2s=",
                        "path": "github.com/stretchr/testify/assert",
                        "revision": "2402e8e7a02fc811447d11f881aa9746cdc57983",
                        "revisionTime": "2016-12-17T20:04:45Z"
                },
                {
-                       "checksumSHA1": "fg3TzS9/QK3wZbzei3Z6O8XPLHg=",
+                       "checksumSHA1": "D4bXPof9qSyXm4ZssBn6J5/FX0U=",
                        "path": "github.com/stretchr/testify/http",
                        "revision": "2402e8e7a02fc811447d11f881aa9746cdc57983",
                        "revisionTime": "2016-12-17T20:04:45Z"
                },
                {
-                       "checksumSHA1": "WGwMB6WljaeGKzuNCX5M4E8LADE=",
+                       "checksumSHA1": "wjNVpZHpkT96ygOH5xN4rDfZdjE=",
                        "path": "github.com/stretchr/testify/mock",
                        "revision": "2402e8e7a02fc811447d11f881aa9746cdc57983",
                        "revisionTime": "2016-12-17T20:04:45Z"
                },
                {
-                       "checksumSHA1": "2PpOCNkWnshDrXeCVH2kp3VHhIM=",
+                       "checksumSHA1": "oqDuQM/+8vHvSt7f8w7WJfmHqws=",
                        "path": "github.com/stretchr/testify/require",
                        "revision": "2402e8e7a02fc811447d11f881aa9746cdc57983",
                        "revisionTime": "2016-12-17T20:04:45Z"
                },
                {
-                       "checksumSHA1": "Dp8TnT65nINpRF/PrxrgmnYLsyA=",
+                       "checksumSHA1": "NtqQXT/XuYBUDnOCC0kn4SXilkA=",
                        "path": "github.com/stretchr/testify/suite",
                        "revision": "2402e8e7a02fc811447d11f881aa9746cdc57983",
                        "revisionTime": "2016-12-17T20:04:45Z"
                },
                {
-                       "checksumSHA1": "pTftHZcNvwQYurltTyqMjG3z2MM=",
+                       "checksumSHA1": "y1pr0JaeOr9y7DQrjYkYpncnjl4=",
                        "path": "github.com/tus/tusd",
                        "revision": "f4810ff42954f4de27e3e4bb3f0ff730aa78e9ab",
                        "revisionTime": "2016-12-20T16:25:24Z"
                },
                {
-                       "checksumSHA1": "sLEH5IAja1Dp4BP2bLnBTywrKdw=",
+                       "checksumSHA1": "vYTCzxDT2HriptOYuyxt9ydfNuI=",
                        "path": "github.com/tus/tusd/filestore",
                        "revision": "f4810ff42954f4de27e3e4bb3f0ff730aa78e9ab",
                        "revisionTime": "2016-12-20T16:25:24Z"
                },
                {
-                       "checksumSHA1": "PV51jijc4Ucfn3IoVSVnY//Vlzg=",
+                       "checksumSHA1": "SoEo838RdEatCmQoaMxUbvZhcZ8=",
                        "path": "github.com/tus/tusd/uid",
                        "revision": "f4810ff42954f4de27e3e4bb3f0ff730aa78e9ab",
                        "revisionTime": "2016-12-20T16:25:24Z"
@@ -165,7 +165,7 @@
                        "revisionTime": "2016-12-14T18:38:57Z"
                },
                {
-                       "checksumSHA1": "0zBf3a7Gzqypt1sWefwx4NkOEIc=",
+                       "checksumSHA1": "kopX8ny86JsJBgQe65Tz/5cf838=",
                        "path": "gopkg.in/Acconut/lockfile.v1",
                        "revision": "941fc61ce4d67dd432b94ed0bd445ad5fb9391a9",
                        "revisionTime": "2016-05-20T15:30:20Z"
warning: LF will be replaced by CRLF in vendor/vendor.json.
```

**Comment by kaey** [comment]
@alexandrevicenzi Thats a git for windows problem, not govendor's. See #255.

**Comment by alexandrevicenzi** [comment]
@kaey the problem is that `govendor sync` changes the vendor.json, that's not right.

**Comment by kardianos** [comment]
@alexandrevicenzi I think both you and @kaey could be correct.

@kaey What do you think if the behavior of `govendor sync` was modified to print warnings and exit != 0 if the checksum didn't match, and not to write to the vendor.json file?

**Comment by bep** [comment]
> was modified to print warnings

Please make it a big and bold ERROR -- people use this to issue bugfix releases, and thinking the `govendor sync` was successful can be pretty bad.

**Comment by moorereason** [comment]
Yes, throw a big error and do not write to the vendor.json (unless `-force` or something).

However, my other recommendation would be to  **remove the checksum altogether**.  See #129.

The only problems we've had with govendor involve these file content checksums.  I haven't looked into how they're used internally (so I'm probably ignorant of why they're essential), but I doubt that the checksums are worth the hassle.  No other package managers I could find compare file checksums.  The VCS does this already, so let it do it's job.

**Comment by kardianos** [comment]
@moorereason I'm aware of the issue of git on windows' default setting where by it doesn't preserve file integrity. One way to possibly fix that in govendor is to ignore "\r" bytes in ".go" files, and might solve the remaining problems here.

Right now `govendor status` uses the checksums to determine if any files were added, removed, modified from their prior state. Sometimes this happens by accident, sometimes on purpose. The reason govendor updates the hash on sync is it *does* trust the git checksums. However, I can see how this can cause problems in certain workflows, so I'd like to fix that.

But I'd also like to keep the ability to record when a file has changed from the original checksum. That is an important feature to me. I'd be willing to omit any "\r" bytes in the checksum.

**Comment by kaey** [comment]
> What do you think if the behavior of govendor sync was modified to print warnings and exit != 0 if the checksum didn't match, and not to write to the vendor.json file?
Suppose it does that, how would end user fix that? What if developers use both nix and windows?

> I'd be willing to omit any "\r" bytes in the checksum.
SGTM.

**Comment by nkovacs** [comment]
> One way to possibly fix that in govendor is to ignore "\r" bytes in ".go" files, and might solve the remaining problems here.

You include all the files in the hash, don't you? You'll have to ignore "\r" in all files that git considers text files for this to work, or a readme.md will also cause the hash to change.
Or you could ignore "\r" in all files, including binary files, but that would lead to false negatives.

**Comment by galopin** [comment]
@kardianos IMHO ignoring "\r" chars on all platforms might be a mistake. The ouput of the above command would let you make an informed decision. When `<eolinfo>` in the index (`i/<eolinfo>`) and in the working tree (`w/<eolinfo>`) do not match, replacing all occurences of `w/<eolinfo>` with `i/<eolinfo>` would allow you to compute a consistent SHA1 checksum…

**Comment by faide** [comment]
For the record I am using mercurial as my DVCS on an [opensource project](https://bitbucket.org/orus-io/xbus) and I still have the same issue.

Development occurs on Linux and example clone below is performed on Linux

```bash

$ cd $GOPATH/src
$ mkdir -p bitbucket.org/orus-io
$ cd bitbucket.org/orus-io
$ hg clone https://bitbucket.org/orus-io/xbus
$ cd xbus
$ govendor sync
$ hg st
M vendor/vendor.json # <-- vendor file was modified...
$ hg diff
```
```diff
diff --git a/vendor/vendor.json b/vendor/vendor.json
--- a/vendor/vendor.json
+++ b/vendor/vendor.json
@@ -3,7 +3,7 @@
        "ignore": "test",
        "package": [
                {
-                      "checksumSHA1": "mh8NkqMO/s8UP/Pi8xL5bqbWP3Y=",
+                      "checksumSHA1": "Wz6vEtMf5uGVn7+JvmymjdDFfEA=",
                        "path": "bitbucket.org/orus-io/go-xbus",
                        "revision": "4b06fefa3000",
                        "revisionTime": "2017-12-11T09:53:21Z",
@@ -162,7 +162,7 @@
                        "revisionTime": "2016-04-26T17:35:18Z"
                },
                {
-                      "checksumSHA1": "KzsJ/GkPEyvxG8eBcKrHRFoPhzU=",
+                      "checksumSHA1": "xppHi82MLqVx1eyQmbhTesAEjx8=",
                        "path": "github.com/lib/pq/oid",
                        "revision": "dd3290b2f71a8b30bee8e4e75a337a825263d26f",
                        "revisionTime": "2016-04-26T17:35:18Z"

... diff output abridged ...

```

**Comment by opsramp** [comment]
@kardianos ,
Any update on this issue? I am using my code both in windows and linux. How can i overcome this checksum change issue?

**Comment by kardianos** [comment]
This is a valid issue. I haven't had time to address it, nor are there PRs which do.

If this is an issue for you, my recommendation is start planning to migrate to vgo `go get -u golang.org/x/vgo`. It supports the `sync` use case much better is most likely the future of dependency management in go.
